### DISCUSSION TITLE
Proposal: treat methods as delegates

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#5062] [Language Ideas]
## Proposal: treat methods as delegates

```csharp
static int TestMethod(int x) => x;
static void Main(string[] args)
{
    var result = TestMethod.Invoke(3); // same with TestMethod(3), TestMethod is Func<int, int>
}
```

With this proposal, we can extend the method itself by extension methods. For example:
```csharp
public static class FuncExtensions
{
    public static Func<T, Func<U, V>> Currying<T, U, V>(this Func<T, U, V> f) => x => y => f(x, y);
}

class Program
{
    static int Test(int a, int b) => a + b;
    static Main(string[] args)
    {
        Test.Currying()(4)(6); //10
    }
}
```

and there will be no need to create a Func:
```csharp
public static class FuncExtensions
{
    public static Func<T, Func<U, V>> Currying<T, U, V>(this Func<T, U, V> f) => x => y => f(x, y);
}

class Program
{
    static int Test(int a, int b) => a + b;
    static Main(string[] args)
    {
        new Func<int, int, int>((x, y) => Test(x, y)).Currying()(4)(6); // it's redundant
    }
}
```

**Comment by HaloFour** [comment]
What problem is this trying to solve?  How would you deal with overloads?

**Comment by hez2010** [comment]
@HaloFour It's especially useful for functional programming in C#.
Although we have F#, but if we can treat functions/methods in C# as delegates (Func, Action), it will be better. And also it can reduce lots of redundant codes about creating delegates from methods/functions, and allows us to extend existing functions and methods using extension methods.

**Comment by HaloFour** [comment]
But what problems does that solve?  You can create and use delegates pretty easily.  You can pass method groups as arguments to methods directly, as long as the delegate type can be resolved unambiguously.  And conversion through a delegate adds overhead both in allocations and during invocations.

And you'll have to solve for the overload problem.  Functional languages don't offer overloads, at least in the same contexts where you can use currying or application.

**Comment by Tom01098** [comment]
I love currying and partial application in F#, but it really isn't suited to the way that C#s model of programming works. I think that it pretty much only works with the calling syntax in functional languages, and might cause havoc with overloading as mentioned above.

**Comment by ymassad** [comment]
Related: #1943

**Comment by hez2010** [comment]
I think this is partially covered by lambda improvements in C# 10. 

```csharp
var f = TestMethod;
f(3);
```

But still we can not write this to use the method directly as a delegate:

```csharp
TestMethod.Invoke(3);
```
